### ISSUE TITLE
Allow passing generic type to `initDataset()`

### ISSUE THREAD

**Issue by mongodben** [braintrustdata/braintrust-sdk#384]
The function `initDataset()` doesn't return typed values. this makes it burdensome to use within projects. 

you should be able to pass the type of your dataset, probably conforming to some base interface, to the function. something like:

```ts
type initDataset<T extends DatasetItem> = (arg) => (...do stuff w/ DatasetItem)

```

**Comment by ankrgyl** [comment]
We've been adding optional zod schemas to a bunch of new surface area -- perhaps the right thing to do here would be to let you pass in a schema that gives you type safety and can be validated at runtime?

**Comment by mongodben** [comment]
> We've been adding optional zod schemas to a bunch of new surface area -- perhaps the right thing to do here would be to let you pass in a schema that gives you type safety and can be validated at runtime?

yeah that makes lots of sense to me. that's actually what we're doing here https://github.com/mongodb/chatbot/pull/505. and then the typing could be inferred internally from the schema with `z.infer<typeof ZodSchema>`

**Comment by mongodben** [comment]
related to this, it'd be nice to allow better tying on dataset entries using generics. right now it's this:

https://github.com/braintrustdata/braintrust-sdk/blob/f14bcf0c9d9d4441837089d9406a275ec7bae489/core/js/src/object.ts#L145

```ts
interface NewDatasetRecord {
  id: string;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  input: any;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  expected: any;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  tags: any;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  metadata: any;
}
```

also i think would make sense if the `id` weren't required for the user facing version since i would prefer braintrust to handle under the hood.

smthn more like:


interface NewDatasetRecord<Input, Expected, Tags extends string[], metadata extends Record<string, unknown>> {
  id?: string;
  input: Input;
  expected: Expected;
  tags: Tags;
  metadata: Metadata;
}
